### ISSUE TITLE
Middleware should catch all requests, even if handler URI is not registered

### ISSUE THREAD

**Issue by habib535** [Starcounter/Home#154]
Technically there is no difference between the following two URLs

1. `http://127.0.0.1:8080/Search`
2. `http://127.0.0.1:8080/Search/`

But the second one will return a 404 because in code we have registered only
`Handle.GET("/Search"`

I added the following Middleware code so the second URL can be mapped to the first one but the Middleware code is not executing either
```
            Application.Current.Use((Request request) =>
            {
                request.Uri = request.Uri.TrimEnd('\\');
                return Self.GET(request.Uri);
            });
```
Do we have any place where I can add my filter logic for any kind of request?

**Comment by Mackiovello** [comment]
What is the order of your middleware?

**Comment by habib535** [comment]
Middleware code is placed above all `Handle.Get` mathods

**Comment by miyconst** [comment]
>Technically there is no difference between the following two URLs

This is wrong. Two URLs with and without trailing `/` are different values. Server should not respond with the same response to `http://127.0.0.1:8080/Search` and `http://127.0.0.1:8080/Search/` since those two are semantically different URLs.

If `/Search/` should point to `/Search` then a redirect should happen.

--

Still, your question is valid. I suppose that middlewares should catch **all** requests regardless of whether it's a registered handler or not.

cc @alemoi

**Comment by Mackiovello** [comment]
Do you have any other middleware that is registered before your middleware? That could make it so that your middleware is never called.

**Comment by habib535** [comment]
Ah! got your point now. No, this is the only Middleware

**Comment by miyconst** [comment]
@per-samuelsson, as far as I know, current middleware was implemented by you. Could you take a look at the issue please?

**Comment by per-samuelsson** [comment]
> @per-samuelsson, as far as I know, current middleware was implemented by you. 

Actually, what's underneath is still request/response filtering just as it were before. I suggested and implemented the reshaping of how we surface them, but never touched filtering itself.

> Could you take a look at the issue please?

I still can, yeah.

**Comment by per-samuelsson** [comment]
So far, from what I can see, this seem to be by design.

Somewhere in the host, a set of built-in handlers are registered. One of them is **"GET /{?}"**. This handler is registered with `HandlerOptions.SkipRequestFilters = true` and hence, our REST server implementation skip running registered middleware / filters.

Now, since `http://127.0.0.1:8080/Search/` would not match any registered handler, it hits the built-in one I mentioned above, and hence request filtering is disabled.

Why it's designed like that, I have no clue. Can't even tell if it's "right" or "wrong". Maybe @alemoi can be of help?

Also, @miyconst please advice if I should try diagnose further.

**Comment by miyconst** [comment]
@per-samuelsson thank you for the checkup, let's wait for @alemoi and see what he says.

**Comment by linkdata** [comment]
What @miyconst said. They're different requests. Usually `path/` will try to get `path/index.html` on other web servers. Usually configurable what it looks for. E.g. http://nginx.org/en/docs/http/ngx_http_index_module.html

**Comment by miyconst** [comment]
@linkdata this issue is not to discuss whether `/xxx` and `/xxx/` should end up in the same handler, but whether request filters should catch all incoming requests regardless of whether someone registered such handler or not.

**Comment by linkdata** [comment]
Oh, sorry, just posted here following a link from slack.

**Comment by alemoi** [comment]
> @per-samuelsson thank you for the checkup, let's wait for @alemoi and see what he says.

I have tested removal of `SkipRequestFilters = true` from static file resource handlers and all seemed to work fine. I tested with People, Images, SignIn, Launchpad apps. Maybe I need to check some other apps that are actually doing something with filters? CC @warpech 
Not pushing anything to 2.3 yet, but the branch is `v2.3-static-filters`

**Comment by warpech** [comment]
I am making a [build](http://teamcity.starcounter.org/viewLog.html?buildId=38842&tab=buildResultsDiv&buildTypeId=Starcounter_V23NightlyWindows) to test this in Web Platform Team projects:

- CompositionProvider
- KitchenSink
- Website

**Comment by warpech** [comment]
@alemoi could you pls merge current `v2.3` to `v2.3-static-filters`?

Website project build fails with an error:

> 
    Api\MainHandlers.cs(83, 29): error CS0117: 'Session' does not contain a definition for 'RunTask'
    ViewModels\SettingsPage.json.cs(84, 29): error CS1061: 'Session' does not contain a definition for 'RunTask' and no extension method 'RunTask' accepting a first argument of type 'Session' could be found (are you missing a using directive or an assembly reference?)
    ViewModels\SettingsPage.json.cs(120, 29): error CS1061: 'Session' does not contain a definition for 'RunTask' and no extension method 'RunTask' accepting a first argument of type 'Session' could be found (are you missing a using directive or an assembly reference?)

**Comment by alemoi** [comment]
> @alemoi could you pls merge current v2.3 to v2.3-static-filters?

Hmm, didn't understand, sorry. How its related to this issue? And are you sure that v2.3 to v2.3-static-filters? Not wise versa?
`

**Comment by warpech** [comment]
> Hmm, didn't understand, sorry. How its related to this issue? And are you sure that v2.3 to v2.3-static-filters? Not wise versa?

Website app does not build because it requires an API that is available in `v2.3` but not in `v2.3-static-filters`. That's why I am asking to merge the current `v2.3` to `v2.3-static-filters` so I can test Website app.

**Comment by alemoi** [comment]
@warpech done! Now Level1 `v2.3-static-filters` contains the latest changes from `v2.3`.

**Comment by warpech** [comment]
- KitchenSink [passed](http://teamcity.starcounter.org/viewLog.html?buildId=39067&tab=buildResultsDiv&buildTypeId=Apps_KitchenSink_Develop) ✔️
- Website [passed](http://teamcity.starcounter.org/viewLog.html?buildId=39068&tab=buildResultsDiv&buildTypeId=Apps_Website_Develop) ✔️
- CompositionProvider [passed](http://teamcity.starcounter.org/viewLog.html?buildId=39069&tab=buildResultsDiv&buildTypeId=Apps_CompositionProvider_Develop) ✔️

So I approve this change happily.

**Notice**: this change was scheduled in Odessa and you made it in `v2.3`. Is this on purpose? I don't see a strong reason to release it in 2.3.1

**Comment by miyconst** [comment]
I don't mind getting it in `2.3` at all.

**Comment by alemoi** [comment]
The fix is pushed to v2.3 and develop. Thanks.
